src/Orleans/Threading/CancellationTokenWrapper.cs

```diff
@@ -0,0 +1,80 @@
+﻿using System;
+using System.Threading;
+using Orleans.CodeGeneration;
+using Orleans.Runtime;
+using Orleans.Serialization;
+
+namespace Orleans.Threading
+{
+    /// <summary>
+    /// Used as replacement of CancellationToken during network roundtrips
+    /// </summary>
+    [Serializable]
+    internal class CancellationTokenWrapper
+    {
+        public CancellationTokenWrapper(Guid id, CancellationToken cancellationToken, GrainReference target)
+            : this(id, cancellationToken)
+        {
+            TargetGrainReference = target;
+        }
+
+        public CancellationTokenWrapper(Guid id, CancellationToken cancellationToken)
+        {
+            Id = id;
+            CancellationToken = cancellationToken;
+            WentThroughSerialization = false;
+        }
+
+        /// <summary>
+        /// Unique id of concrete token
+        /// </summary>
+        public Guid Id { get; private set; }
+
+        /// <summary>
+        /// Original request target grain reference
+        /// </summary>
+        public GrainReference TargetGrainReference { get; private set; }
+
+        /// <summary>
+        /// Cancellation token
+        /// </summary>
+        public CancellationToken CancellationToken { get; private set; }
```

Comment:
So need to add [DoNotSerialize]. Even if we serialize ourselves, its better to be declarative for binary ser. Everythijng has to work with binary ser as well, right? And json.